From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Kristof Imeri <kristof@lanth.me>
Date: Sun, 18 Oct 2026 04:30:30 +0200
Subject: [PATCH 36/50] clutter: draw text and offscreen textures at the view's
 own scale on whole pixels

On a view with a fractional scale such as 1.5, resources were made at the
scale rounded up to 2 and shrunk by linear filtering when painted, so text
and icons came out soft, and softer still wherever an actor's position fell
between device pixels, as it does for most popups and labels.

Lay text out and size offscreen effects at the actor's unrounded resource
scale, and move text layouts and textures that map one texel to one device
pixel onto whole pixels when they are painted. Scaled or rotated actors,
such as those in an animation that zooms, keep their exact position.
---
 clutter/clutter/clutter-actor-private.h    |  2 -
 clutter/clutter/clutter-actor.c            | 18 +++++++-
 clutter/clutter/clutter-actor.h            |  3 ++
 clutter/clutter/clutter-offscreen-effect.c | 11 +++--
 clutter/clutter/clutter-paint-nodes.c      | 51 ++++++++++++++++++++++
 clutter/clutter/clutter-paint-nodes.h      |  4 ++
 clutter/clutter/clutter-private.h          |  7 +++
 clutter/clutter/clutter-util.c             | 39 +++++++++++++++++
 clutter/clutter/pango/clutter-text.c       | 42 ++++++++++--------
 9 files changed, 151 insertions(+), 26 deletions(-)

diff --git a/clutter/clutter/clutter-actor-private.h b/clutter/clutter/clutter-actor-private.h
index 8defef325d8eb4add88258f134aa863c35151441..4fc67040c63d8b8e6c9ee9702698a39d0b649fcb 100644
--- a/clutter/clutter/clutter-actor-private.h
+++ b/clutter/clutter/clutter-actor-private.h
@@ -277,8 +277,6 @@ gboolean                        clutter_actor_is_unrealized_mapped
 void                            clutter_actor_clear_stage_views_recursive               (ClutterActor *actor,
                                                                                          gboolean      stop_transitions);
 
-float                           clutter_actor_get_real_resource_scale                   (ClutterActor *actor);
-
 void clutter_actor_finish_layout (ClutterActor *self,
                                   int           phase);
 
diff --git a/clutter/clutter/clutter-actor.c b/clutter/clutter/clutter-actor.c
index 49c98833dfd1cdb6cd48c594aeb67f82860d7709..6af7b0340849abd7cbeb5902af1b3ec1e8c31dd4 100644
--- a/clutter/clutter/clutter-actor.c
+++ b/clutter/clutter/clutter-actor.c
@@ -15084,6 +15084,19 @@ clutter_actor_clear_stage_views_recursive (ClutterActor *self,
                            NULL);
 }
 
+/**
+ * clutter_actor_get_real_resource_scale:
+ * @self: A #ClutterActor
+ *
+ * Retrieves the scale of the highest scaled #ClutterStageView the actor
+ * gets painted on, without rounding it up to a whole number.
+ *
+ * Resources made at this scale map one to one onto the pixels of a view
+ * with a fractional scale. The "real-resource-scale-changed" signal tells
+ * when it changes.
+ *
+ * Returns: The unrounded resource scale of the actor
+ */
 float
 clutter_actor_get_real_resource_scale (ClutterActor *self)
 {
@@ -18929,7 +18942,10 @@ clutter_actor_create_texture_paint_node (ClutterActor *self,
   clutter_paint_node_set_static_name (node, "Texture");
 
   if (priv->content_repeat == CLUTTER_REPEAT_NONE)
-    clutter_paint_node_add_rectangle (node, &box);
+    {
+      clutter_pipeline_node_set_whole_pixels (CLUTTER_PIPELINE_NODE (node), TRUE);
+      clutter_paint_node_add_rectangle (node, &box);
+    }
   else
     {
       float t_w = 1.f, t_h = 1.f;
diff --git a/clutter/clutter/clutter-actor.h b/clutter/clutter/clutter-actor.h
index 2a63c1f76b9571d1faebad4f1ebe95e5b9e91e00..4bda55062d8fdecffe021e7081bc6e566b7162dc 100644
--- a/clutter/clutter/clutter-actor.h
+++ b/clutter/clutter/clutter-actor.h
@@ -555,6 +555,9 @@ gboolean                        clutter_actor_get_paint_box
 CLUTTER_EXPORT
 float                           clutter_actor_get_resource_scale                (ClutterActor *self);
 
+CLUTTER_EXPORT
+float                           clutter_actor_get_real_resource_scale           (ClutterActor *self);
+
 CLUTTER_EXPORT
 gboolean                        clutter_actor_has_overlaps                      (ClutterActor               *self);
 
diff --git a/clutter/clutter/clutter-offscreen-effect.c b/clutter/clutter/clutter-offscreen-effect.c
index 0f3fd83e62aeafb09e218a835efcbed0a0cdbc60..60675fce9ea7659056874456f1eb9a98af730342 100644
--- a/clutter/clutter/clutter-offscreen-effect.c
+++ b/clutter/clutter/clutter-offscreen-effect.c
@@ -310,7 +310,6 @@ clutter_offscreen_effect_pre_paint (ClutterEffect       *effect,
   gfloat stage_width, stage_height;
   gfloat target_width = -1, target_height = -1;
   float resource_scale;
-  float ceiled_resource_scale;
 
   if (!clutter_actor_meta_get_enabled (CLUTTER_ACTOR_META (effect)))
     goto disable_effect;
@@ -323,9 +322,8 @@ clutter_offscreen_effect_pre_paint (ClutterEffect       *effect,
 
   resource_scale = clutter_actor_get_real_resource_scale (priv->actor);
 
-  ceiled_resource_scale = ceilf (resource_scale);
-  stage_width *= ceiled_resource_scale;
-  stage_height *= ceiled_resource_scale;
+  stage_width *= resource_scale;
+  stage_height *= resource_scale;
 
   /* Get the minimal bounding box for what we want to paint, relative to the
    * parent of priv->actor. Note that we may actually be painting a clone of
@@ -358,7 +356,7 @@ clutter_offscreen_effect_pre_paint (ClutterEffect       *effect,
       priv->fbo_offset_y = (int) (box.y1 - raw_box.y1);
     }
 
-  clutter_actor_box_scale (&box, ceiled_resource_scale);
+  clutter_actor_box_scale (&box, resource_scale);
   clutter_actor_box_get_size (&box, &target_width, &target_height);
 
   target_width = ceilf (target_width);
@@ -430,6 +428,7 @@ clutter_offscreen_effect_real_paint_target (ClutterOffscreenEffect *effect,
   cogl_pipeline_set_color (priv->pipeline, &color);
 
   pipeline_node = clutter_pipeline_node_new (priv->pipeline);
+  clutter_pipeline_node_set_whole_pixels (CLUTTER_PIPELINE_NODE (pipeline_node), TRUE);
   clutter_paint_node_set_static_name (pipeline_node,
                                       G_OBJECT_TYPE_NAME (effect));
   clutter_paint_node_add_child (node, pipeline_node);
@@ -459,7 +458,7 @@ clutter_offscreen_effect_paint_texture (ClutterOffscreenEffect *effect,
   graphene_matrix_t transform;
   float unscale;
 
-  unscale = 1.0f / clutter_actor_get_resource_scale (priv->actor);
+  unscale = 1.0f / clutter_actor_get_real_resource_scale (priv->actor);
   graphene_matrix_init_scale (&transform, unscale, unscale, 1.0f);
   graphene_matrix_translate (&transform,
                              &GRAPHENE_POINT3D_INIT (priv->fbo_offset_x,
diff --git a/clutter/clutter/clutter-paint-nodes.c b/clutter/clutter/clutter-paint-nodes.c
index 989e96d85b6c014d0a71fd50f368fafb53918fab..1d45293f7a155ee84a6cb112b24943212064be64 100644
--- a/clutter/clutter/clutter-paint-nodes.c
+++ b/clutter/clutter/clutter-paint-nodes.c
@@ -408,6 +408,7 @@ struct _ClutterPipelineNode
   ClutterPaintNode parent_instance;
 
   CoglPipeline *pipeline;
+  gboolean whole_pixels;
 };
 
 /**
@@ -459,6 +460,34 @@ get_target_framebuffer (ClutterPaintNode    *node,
   return clutter_paint_context_get_framebuffer (paint_context);
 }
 
+static void
+draw_texture_on_whole_pixels (CoglFramebuffer *fb,
+                              CoglPipeline    *pipeline,
+                              const float     *rect)
+{
+  CoglTexture *texture = cogl_pipeline_get_layer_texture (pipeline, 0);
+  float offset_x, offset_y;
+  gboolean aligned;
+
+  aligned = texture && rect[2] != rect[0] &&
+    _clutter_util_offset_to_whole_pixel (fb, rect[0], rect[1],
+                                         cogl_texture_get_width (texture) *
+                                         (rect[6] - rect[4]) / (rect[2] - rect[0]),
+                                         &offset_x, &offset_y);
+  if (aligned)
+    {
+      cogl_framebuffer_push_matrix (fb);
+      cogl_framebuffer_translate (fb, offset_x, offset_y, 0.0f);
+    }
+
+  cogl_framebuffer_draw_textured_rectangle (fb, pipeline,
+                                            rect[0], rect[1], rect[2], rect[3],
+                                            rect[4], rect[5], rect[6], rect[7]);
+
+  if (aligned)
+    cogl_framebuffer_pop_matrix (fb);
+}
+
 static void
 clutter_pipeline_node_draw (ClutterPaintNode    *node,
                             ClutterPaintContext *paint_context)
@@ -501,6 +530,11 @@ clutter_pipeline_node_draw (ClutterPaintNode    *node,
           break;
 
         case PAINT_OP_TEX_RECT:
+          if (pnode->whole_pixels)
+            {
+              draw_texture_on_whole_pixels (fb, pnode->pipeline, op->op.texrect);
+              break;
+            }
           cogl_framebuffer_draw_textured_rectangle (fb,
                                                     pnode->pipeline,
                                                     op->op.texrect[0],
@@ -591,6 +625,23 @@ clutter_pipeline_node_new (CoglPipeline *pipeline)
   return (ClutterPaintNode *) res;
 }
 
+/**
+ * clutter_pipeline_node_set_whole_pixels:
+ * @node: a #ClutterPipelineNode
+ * @whole_pixels: whether to draw on whole pixels
+ *
+ * Moves textured rectangles that map one texel to one pixel of the
+ * framebuffer onto whole pixels, so they are drawn without resampling.
+ */
+void
+clutter_pipeline_node_set_whole_pixels (ClutterPipelineNode *node,
+                                        gboolean             whole_pixels)
+{
+  g_return_if_fail (CLUTTER_IS_PIPELINE_NODE (node));
+
+  node->whole_pixels = whole_pixels;
+}
+
 /**
  * ClutterColorNode:
  */
diff --git a/clutter/clutter/clutter-paint-nodes.h b/clutter/clutter/clutter-paint-nodes.h
index 98642afe9509c6643c9d60951355b302f5241ba1..26eb298124c1659ebefb2a0bafa57baf11cb07b5 100644
--- a/clutter/clutter/clutter-paint-nodes.h
+++ b/clutter/clutter/clutter-paint-nodes.h
@@ -88,6 +88,10 @@ GType clutter_pipeline_node_get_type (void);
 CLUTTER_EXPORT
 ClutterPaintNode *      clutter_pipeline_node_new       (CoglPipeline          *pipeline);
 
+CLUTTER_EXPORT
+void                    clutter_pipeline_node_set_whole_pixels (ClutterPipelineNode *node,
+                                                                gboolean             whole_pixels);
+
 #define CLUTTER_TYPE_ACTOR_NODE                 (clutter_actor_node_get_type ())
 #define CLUTTER_ACTOR_NODE(obj)                 (G_TYPE_CHECK_INSTANCE_CAST ((obj), CLUTTER_TYPE_ACTOR_NODE, ClutterActorNode))
 #define CLUTTER_IS_ACTOR_NODE(obj)              (G_TYPE_CHECK_INSTANCE_TYPE ((obj), CLUTTER_TYPE_ACTOR_NODE))
diff --git a/clutter/clutter/clutter-private.h b/clutter/clutter/clutter-private.h
index 418a52e9c77b929463712870a4d4ac55e6628a12..942b4b122159252e7dba55126ca8fd968b08a888 100644
--- a/clutter/clutter/clutter-private.h
+++ b/clutter/clutter/clutter-private.h
@@ -120,6 +120,13 @@ void  _clutter_util_fully_transform_vertices (const graphene_matrix_t  *modelvie
                                               graphene_point3d_t       *vertices_out,
                                               int                       n_vertices);
 
+gboolean _clutter_util_offset_to_whole_pixel (CoglFramebuffer *framebuffer,
+                                              float            x,
+                                              float            y,
+                                              float            pixels_per_unit,
+                                              float           *offset_x,
+                                              float           *offset_y);
+
 typedef enum _ClutterCullResult
 {
   CLUTTER_CULL_RESULT_IN,
diff --git a/clutter/clutter/clutter-util.c b/clutter/clutter/clutter-util.c
index d017006f20aa7a631214d4313ef450ae4191ead3..bfa6284fbd03323188af9868116467fc7eb0bb58 100644
--- a/clutter/clutter/clutter-util.c
+++ b/clutter/clutter/clutter-util.c
@@ -108,6 +108,45 @@ _clutter_util_fully_transform_vertices (const graphene_matrix_t  *modelview,
     }
 }
 
+#define PIXEL_EPSILON (1.f / 128.f)
+
+gboolean
+_clutter_util_offset_to_whole_pixel (CoglFramebuffer *framebuffer,
+                                     float            x,
+                                     float            y,
+                                     float            pixels_per_unit,
+                                     float           *offset_x,
+                                     float           *offset_y)
+{
+  graphene_matrix_t modelview, projection;
+  float viewport[4];
+  graphene_point3d_t in[3] = {
+    GRAPHENE_POINT3D_INIT (x, y, 0.f),
+    GRAPHENE_POINT3D_INIT (x + 1.f, y, 0.f),
+    GRAPHENE_POINT3D_INIT (x, y + 1.f, 0.f),
+  };
+  graphene_point3d_t out[3];
+  float step_x, step_y;
+
+  cogl_framebuffer_get_modelview_matrix (framebuffer, &modelview);
+  cogl_framebuffer_get_projection_matrix (framebuffer, &projection);
+  cogl_framebuffer_get_viewport4fv (framebuffer, viewport);
+  _clutter_util_fully_transform_vertices (&modelview, &projection, viewport,
+                                          in, out, G_N_ELEMENTS (in));
+
+  step_x = out[1].x - out[0].x;
+  step_y = out[2].y - out[0].y;
+  if (fabsf (fabsf (step_x) - pixels_per_unit) > PIXEL_EPSILON ||
+      fabsf (fabsf (step_y) - pixels_per_unit) > PIXEL_EPSILON ||
+      fabsf (out[1].y - out[0].y) > PIXEL_EPSILON ||
+      fabsf (out[2].x - out[0].x) > PIXEL_EPSILON)
+    return FALSE;
+
+  *offset_x = (roundf (out[0].x) - out[0].x) / step_x;
+  *offset_y = (roundf (out[0].y) - out[0].y) / step_y;
+  return TRUE;
+}
+
 typedef struct
 {
   ClutterProgressFunc func;
diff --git a/clutter/clutter/pango/clutter-text.c b/clutter/clutter/pango/clutter-text.c
index c897430f087c551c39957951b064ab027800c8bc..8ce30ddf77730a1201c4188e0c264182471c5c5c 100644
--- a/clutter/clutter/pango/clutter-text.c
+++ b/clutter/clutter/pango/clutter-text.c
@@ -697,7 +697,7 @@ ensure_effective_pango_scale_attributes (ClutterText *self)
   float resource_scale;
   ClutterTextPrivate *priv = clutter_text_get_instance_private (self);
 
-  resource_scale = clutter_actor_get_resource_scale (CLUTTER_ACTOR (self));
+  resource_scale = clutter_actor_get_real_resource_scale (CLUTTER_ACTOR (self));
 
   if (priv->effective_attrs != NULL)
     {
@@ -1246,7 +1246,7 @@ maybe_create_text_layout_with_resource_scale (ClutterText *text,
 {
   float resource_scale;
 
-  resource_scale = clutter_actor_get_resource_scale (CLUTTER_ACTOR (text));
+  resource_scale = clutter_actor_get_real_resource_scale (CLUTTER_ACTOR (text));
 
   return create_text_layout_with_scale (text,
                                         allocation_width,
@@ -1278,7 +1278,7 @@ clutter_text_coords_to_position (ClutterText *self,
   g_return_val_if_fail (CLUTTER_IS_TEXT (self), 0);
 
   priv = clutter_text_get_instance_private (self);
-  resource_scale = clutter_actor_get_resource_scale (CLUTTER_ACTOR (self));
+  resource_scale = clutter_actor_get_real_resource_scale (CLUTTER_ACTOR (self));
 
   /* Take any offset due to scrolling into account, and normalize
    * the coordinates to PangoScale units
@@ -1404,7 +1404,7 @@ clutter_text_position_to_coords (ClutterText *self,
 
   g_return_val_if_fail (CLUTTER_IS_TEXT (self), FALSE);
 
-  resource_scale = clutter_actor_get_resource_scale (CLUTTER_ACTOR (self));
+  resource_scale = clutter_actor_get_real_resource_scale (CLUTTER_ACTOR (self));
 
   ret = clutter_text_position_to_coords_internal (self, position,
                                                   x, y, line_height);
@@ -2512,6 +2512,7 @@ clutter_text_paint (ClutterActor        *self,
   float alloc_width;
   float alloc_height;
   float resource_scale;
+  float pixel_offset_x, pixel_offset_y;
 
   fb = clutter_paint_context_get_framebuffer (paint_context);
 
@@ -2530,7 +2531,7 @@ clutter_text_paint (ClutterActor        *self,
       !clutter_text_should_draw_cursor (text))
     return;
 
-  resource_scale = clutter_actor_get_resource_scale (CLUTTER_ACTOR (self));
+  resource_scale = clutter_actor_get_real_resource_scale (CLUTTER_ACTOR (self));
 
   clutter_actor_box_scale (&alloc, resource_scale);
   clutter_actor_box_get_size (&alloc, &alloc_width, &alloc_height);
@@ -2566,10 +2567,10 @@ clutter_text_paint (ClutterActor        *self,
         }
     }
 
+  cogl_framebuffer_push_matrix (fb);
   if (resource_scale != 1.0f)
     {
       float paint_scale = 1.0f / resource_scale;
-      cogl_framebuffer_push_matrix (fb);
       cogl_framebuffer_scale (fb, paint_scale, paint_scale, 1.0f);
     }
 
@@ -2664,13 +2665,16 @@ clutter_text_paint (ClutterActor        *self,
                            priv->text_color.blue / 255.0f,
                            real_opacity / 255.0f);
 
+  if (_clutter_util_offset_to_whole_pixel (fb, priv->text_x, priv->text_y, 1.0f,
+                                           &pixel_offset_x, &pixel_offset_y))
+    cogl_framebuffer_translate (fb, pixel_offset_x, pixel_offset_y, 0.0f);
+
   clutter_show_layout (context, fb, layout, priv->text_x, priv->text_y, &color,
                        color_state, target_color_state);
 
   selection_paint (text, fb, paint_context);
 
-  if (resource_scale != 1.0f)
-    cogl_framebuffer_pop_matrix (fb);
+  cogl_framebuffer_pop_matrix (fb);
 
   if (clip_set)
     cogl_framebuffer_pop_clip (fb);
@@ -2742,6 +2746,7 @@ clutter_text_get_paint_volume (ClutterActor       *self,
   PangoRectangle ink_rect;
   graphene_point3d_t origin;
   float resource_scale;
+  float pixel;
 
   /* ClutterText uses the logical layout as the natural size of the
      actor. This means that it can sometimes paint outside of its
@@ -2761,23 +2766,24 @@ clutter_text_get_paint_volume (ClutterActor       *self,
   if (!clutter_actor_has_allocation (self))
     return FALSE;
 
-  resource_scale = clutter_actor_get_resource_scale (self);
+  resource_scale = clutter_actor_get_real_resource_scale (self);
 
   clutter_paint_volume_init_from_actor (volume, self);
 
   layout = clutter_text_get_layout (text);
   pango_layout_get_extents (layout, &ink_rect, NULL);
 
-  origin.x = pango_to_logical_pixels (ink_rect.x, resource_scale);
-  origin.y = pango_to_logical_pixels (ink_rect.y, resource_scale);
+  pixel = 1.0f / resource_scale;
+  origin.x = pango_to_logical_pixels (ink_rect.x, resource_scale) - pixel;
+  origin.y = pango_to_logical_pixels (ink_rect.y, resource_scale) - pixel;
   origin.z = 0;
   clutter_paint_volume_set_origin (volume, &origin);
   clutter_paint_volume_set_width (volume,
                                   pango_to_logical_pixels (ink_rect.width,
-                                                           resource_scale));
+                                                           resource_scale) + 2 * pixel);
   clutter_paint_volume_set_height (volume,
                                    pango_to_logical_pixels (ink_rect.height,
-                                                            resource_scale));
+                                                            resource_scale) + 2 * pixel);
 
   /* If the cursor is visible then that will likely be drawn
      outside of the ink rectangle so we should merge that in */
@@ -2811,7 +2817,7 @@ clutter_text_get_preferred_width (ClutterActor *self,
   gfloat layout_width;
   gfloat resource_scale;
 
-  resource_scale = clutter_actor_get_resource_scale (self);
+  resource_scale = clutter_actor_get_real_resource_scale (self);
 
   layout = clutter_text_create_layout (text, -1, -1);
   pango_layout_get_extents (layout, NULL, &logical_rect);
@@ -2868,7 +2874,7 @@ clutter_text_get_preferred_height (ClutterActor *self,
       gfloat layout_height;
       gfloat resource_scale;
 
-      resource_scale = clutter_actor_get_resource_scale (self);
+      resource_scale = clutter_actor_get_real_resource_scale (self);
 
       if (priv->single_line_mode)
         for_width = -1;
@@ -3728,7 +3734,6 @@ clutter_text_class_init (ClutterTextClass *klass)
   actor_class->key_focus_out = clutter_text_key_focus_out;
   actor_class->has_overlaps = clutter_text_has_overlaps;
   actor_class->calculate_resource_scale = clutter_text_calculate_resource_scale;
-  actor_class->resource_scale_changed = clutter_text_resource_scale_changed;
   actor_class->unrealize = clutter_text_unrealize;
 
   /**
@@ -4505,6 +4510,9 @@ clutter_text_init (ClutterText *self)
     g_signal_connect (self, "notify::text-direction",
                       G_CALLBACK (clutter_text_direction_changed_cb),
                       NULL);
+  g_signal_connect (self, "real-resource-scale-changed",
+                    G_CALLBACK (clutter_text_resource_scale_changed),
+                    NULL);
 
   priv->input_focus = clutter_text_input_focus_new (self);
 
@@ -6684,7 +6692,7 @@ clutter_text_get_cursor_rect (ClutterText     *self,
   g_return_if_fail (CLUTTER_IS_TEXT (self));
   g_return_if_fail (rect != NULL);
 
-  inverse_scale = 1.f / clutter_actor_get_resource_scale (CLUTTER_ACTOR (self));
+  inverse_scale = 1.f / clutter_actor_get_real_resource_scale (CLUTTER_ACTOR (self));
   priv = clutter_text_get_instance_private (self);
 
   graphene_rect_scale (&priv->cursor_rect,
